### PR TITLE
feat: better lobby exception handling [MTT-2649]

### PR DESCRIPTION
**Pull request by LPLafontaineB**
### Description

This PR is mostly a cherrypick of #568, which restricted popup messages to only some lobby exceptions and removed an unnecessary call to check if a player was part of a lobby before trying to kick them.

On top of the cherrypick, I replaced the handling of 400 bad request which from my earlier tests happened when we tried to input a badly formatted user-code with ValidationError which is the exception reason we receive now.

### Issue Number(s)

[MTT-2649](https://jira.unity3d.com/browse/MTT-2649) and [MTT-2840](https://jira.unity3d.com/browse/MTT-2840)
### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] JIRA ticket ID is in the PR title or at least one commit message
 - [x] Include the ticket ID number within the body message of the PR to create a hyperlink

### CODE REVIEW

**Review by pdeschain — DISMISSED** [review]
👍

File: Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs
Comment (fernando-cortez) [inline Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs:74]: Why not just cache the Regex pattern?

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs]: done

**Review by fernando-cortez — APPROVED** [review]

File: Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs:27]: what's the jira ticket for this lobby side?
Since they are going to fix this, any reason we're doing this workaround for it here at all?

File: Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs:70]: Should we still log the unhandled ones? Or will they be handled somewhere else?

**Review by SamuelBellomo — DISMISSED** [review]

File: Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs:70]: They are already logged with the stack trace from before the async task before creating the UnityServiceErrorMessage that is published and caught here

File: Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs:27]: It was still in discussion whether it was something they wanted to expose in the SDK or not, so no Jira ticket yet.
I added this here to simply prevent us from getting the ValidationError, figuring that if we know it will fail, there is no point of actually calling the service.
However, I got some new feedback from the Lobby team about this, I'll wait for the conclusion of that conversation before merging this PR.

**Comment by LPLafontaineB** [comment]
Putting on hold temporarily, just waiting for some feedback from the lobby team

File: Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs]: I ended up removing this and just doing sanitization with the full alphabet instead for this specific PR. If we want to do more and prevent the ValidationError, I think it should be a new ticket, following the conclusion of the aforementioned conversation

File: Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs
Comment (pdeschain) [inline Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs:52]: I think that you'd get a miscast exception if OriginalException is something else than LobbyServiceException.

Comment (pdeschain) [inline Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs:52]:
```suggestion
            var exception = error.OriginalException as LobbyServiceException;
```

**Review by pdeschain — CHANGES_REQUESTED** [review]

File: Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs]: Normally it should not happen since we only call HandleLobbyError when error.AffectedService equals UnityServiceErrorMessage.Service.Lobby. If it does, it would mean that the UnityServiceErrorMessage was not properly constructed, right?

**Review by fernando-cortez — COMMENTED** [review]
Can approve once comments are addressed 👍.

**Review by SamuelBellomo — DISMISSED** [review]

File: Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs
Comment (pdeschain) [reply Assets/BossRoom/Scripts/Client/UI/UnityServicesUIHandler.cs]: Well, HandleLobbyError method is presuming that no other UnityServices will ever use UnityServiceErrorMessage or something like that That's a wrong assumption. Exception is the type of OriginalException. Everything else is just a possibility.

**Review by SamuelBellomo — APPROVED** [review]

**Review by pdeschain — DISMISSED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Review by pdeschain — APPROVED** [review]

**Pull request merged**
